### ISSUE TITLE
Crash on entering search term

### ISSUE THREAD

**Issue by uli-on** [maltaisn/another-notes-app#47]
Hi maltaisn!
Since updating to v1.4.1, searching is currently not possible any longer – at least on my device. The crash happens the very moment the first letter is entered into the search field.
I hope the log excerpt caught by the "Scoop" app is of help to you.

App v1.4.1
Android 6.0.1 (with GMS 21.18.16), non-rooted
Webview 93.0.4577.62
Moto G4

```
FATAL EXCEPTION: main
Process: com.maltaisn.notes.sync, PID: 11535
java.lang.StringIndexOutOfBoundsException: length=368; index=-10
	at java.lang.String.charAt(Native Method)
	at com.maltaisn.notes.h.v.f.b()
	at com.maltaisn.notes.h.v.j.d()
	at com.maltaisn.notes.h.v.j.a()
	at com.maltaisn.notes.h.v.j.b()
	at com.maltaisn.notes.ui.search.e.E0()
	at com.maltaisn.notes.ui.search.e.B0()
	at com.maltaisn.notes.ui.search.e$d$a.a()
	at kotlinx.coroutines.z2.i.j$a.o()
	at kotlinx.coroutines.z2.i.j$a.k()
	at kotlinx.coroutines.z2.i.i.C()
	at kotlinx.coroutines.z2.i.i.a()
	at androidx.room.z$a$a$a$a.y()
	at g.e0.j.a.a.m()
	at kotlinx.coroutines.w0.run()
	at android.os.Handler.handleCallback(Handler.java:746)
	at android.os.Handler.dispatchMessage(Handler.java:95)
	at android.os.Looper.loop(Looper.java:148)
	at android.app.ActivityThread.main(ActivityThread.java:5443)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:728)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:618)
```

**Comment by uli-on** [comment]
Whew, I've progressed and can provide you steps to reproduce. 

I've emptied the db via prefs, created a new article and searched without weird things happening. With my db back in place I had crashes again.

So I remembered having a crash once when I searched for some term that started with a minus sign immediately followed by a normal character.


So I just emptied the db again and created a new article, this time with an example of my own tagging syntax that I'm using to bridge the gap in Another Notes: a dot immediately followed by the term that serves as the tag. (Example: `.my-tag`.) 

With that in place in a notice, entering a letter into the search field made the crash happen again.

When fixing this (should you want to), please keep in mind that there might be more characters and combinations that the search functionality doesn't like. Thanks.

**Comment by maltaisn** [comment]
Actually it seems you found two bugs. I just fixed the bug for stacktrace you provided, which occured if you had maximum preview lines set to 1 or 2. The crash with the special chars is something else entirely, I'll look into it later.

**Comment by uli-on** [comment]
I can provide you with a log excerpt for the second crash, this time it needed a little shaking the Scoop app.

```
FATAL EXCEPTION: main
Process: com.maltaisn.notes.sync, PID: 27453
java.lang.StringIndexOutOfBoundsException: length=32; index=-7
	at java.lang.String.charAt(Native Method)
	at com.maltaisn.notes.h.v.f.b()
	at com.maltaisn.notes.h.v.j.d()
	at com.maltaisn.notes.h.v.j.a()
	at com.maltaisn.notes.h.v.j.b()
	at com.maltaisn.notes.ui.search.e.E0()
	at com.maltaisn.notes.ui.search.e.B0()
	at com.maltaisn.notes.ui.search.e$d$a.a()
	at kotlinx.coroutines.z2.i.j$a.o()
	at kotlinx.coroutines.z2.i.j$a.k()
	at kotlinx.coroutines.z2.i.i.C()
	at kotlinx.coroutines.z2.i.i.a()
	at androidx.room.z$a$a$a$a.y()
	at g.e0.j.a.a.m()
	at kotlinx.coroutines.w0.run()
	at android.os.Handler.handleCallback(Handler.java:746)
	at android.os.Handler.dispatchMessage(Handler.java:95)
	at android.os.Looper.loop(Looper.java:148)
	at android.app.ActivityThread.main(ActivityThread.java:5443)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:728)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:618)
```

**Comment by maltaisn** [comment]
That actually looks like the first bug too. But that's fine, I already managed to reproduce it and have the stacktrace.

**Comment by uli-on** [comment]
Grrrrrreat! Thank you!!

**Comment by maltaisn** [comment]
I fixed the other bug that occured when you search with a negative term `-word` without anything else. 

The current implementation doesn't allow special characters (hyphens, periods, etc). This is partly by design, to find as many results as possible, ignoring noise like hyphens. Changing that would require changing the whole search system. You can use labels for tags, however you can't search for multiple labels.

You might be interested to know that there are a few undocumented features in the search, namely you can quote stuff to find exact matches and you can use a minus sign to exclude terms. But quoting still doesn't allow hyphens to be matched.

**Comment by uli-on** [comment]
> You can use labels for tags, however you can't search for multiple labels.

The latter was what I tried to mimic, tried to allow for, yes.

Thanks for your quick fix and the search tips!

**Comment by chocmake** [comment]
Just experienced this on v1.5.1. The moment a single character (just alphanumeric) is entered in the search field the app crashes, consistently. I don't have any error tracing logs to provide, unfortunately.

Can't recall if I'd used the search feature prior without the issue (have only had the app installed for about a week without heavy use) but encountered this bug right after an OS factory reset (LineageOS 20 / Android 13) along with restoring the app via backup, then subsequently within Another Notes App importing a previously exported notes backup.

**Comment by maltaisn** [comment]
Can you install Scoop from F-Droid to get a crash report? You just have to let Scoop run and make the app crash and Scoop will give a stacktrace.

**Comment by chocmake** [comment]
> Can you install Scoop from F-Droid to get a crash report?

Here you go.

```
FATAL EXCEPTION: main
Process: com.maltaisn.notes.sync, PID: 4655
java.lang.IndexOutOfBoundsException: setSpan (-23 ... -22) starts before 0
	at android.text.SpannableStringInternal.checkRange(SpannableStringInternal.java:499)
	at android.text.SpannableStringInternal.setSpan(SpannableStringInternal.java:199)
	at android.text.SpannableStringInternal.setSpan(SpannableStringInternal.java:186)
	at android.text.SpannableString.setSpan(SpannableString.java:60)
	at androidx.activity.o.H(SourceFile:69)
	at n3.h.e(SourceFile:187)
	at androidx.recyclerview.widget.RecyclerView$e.f(Unknown Source:0)
	at androidx.recyclerview.widget.RecyclerView$t.i(SourceFile:1228)
	at androidx.recyclerview.widget.StaggeredGridLayoutManager.H0(SourceFile:4)
	at androidx.recyclerview.widget.StaggeredGridLayoutManager.U0(SourceFile:30)
	at androidx.recyclerview.widget.StaggeredGridLayoutManager.d0(Unknown Source:1)
	at androidx.recyclerview.widget.RecyclerView.r(SourceFile:85)
	at androidx.recyclerview.widget.RecyclerView.p(SourceFile:127)
	at androidx.recyclerview.widget.RecyclerView.m(SourceFile:142)
	at androidx.recyclerview.widget.RecyclerView$a.run(Unknown Source:31)
	at android.view.Choreographer$CallbackRecord.run(Choreographer.java:1231)
	at android.view.Choreographer$CallbackRecord.run(Choreographer.java:1239)
	at android.view.Choreographer.doCallbacks(Choreographer.java:899)
	at android.view.Choreographer.doFrame(Choreographer.java:827)
	at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:1214)
	at android.os.Handler.handleCallback(Handler.java:942)
	at android.os.Handler.dispatchMessage(Handler.java:99)
	at android.os.Looper.loopOnce(Looper.java:201)
	at android.os.Looper.loop(Looper.java:288)
	at android.app.ActivityThread.main(ActivityThread.java:7918)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:936)
```

**Comment by chocmake** [comment]
Missed part of the above where it mentioned a bug fixed earlier if settings had less than two maximum lines preview. In my case I have `0` lines preview for the List view mode so tested with 1+ and the app no longer crashed.

If there needs to be visible strings when matching note body text for searches (rather than just showing the item but not showing any body text) then maybe the app could override the setting and display a line.

**Comment by maltaisn** [comment]
Ah I see, I can reproduce it.

**Comment by maltaisn** [comment]
Fixed in v1.5.2
